src/SDAO.t.sol

```diff
@@ -0,0 +1,509 @@
+    }
+
+    function testMetadataFuzz(string calldata name, string calldata symbol) public {
+        SDAO tkn = new SDAO(name, symbol);
+        assertEq(tkn.name(), name);
+        assertEq(tkn.symbol(), symbol);
+        assertEq(tkn.decimals(), 18);
+    }
+
+    function testMintFuzz(address to, uint256 amount) public {
+        vm.assume(to != address(0) && to != address(token));
+
+        token.mint(to, amount);
+
+        assertEq(token.totalSupply(), amount);
+        assertEq(token.balanceOf(to), amount);
+    }
+
+    function testBurnFuzz(address from, uint256 mintAmount, uint256 burnAmount) public {
+        vm.assume(from != address(0) && from != address(token));
+        burnAmount = bound(burnAmount, 0, mintAmount);
+
+        token.mint(from, mintAmount);
+        vm.prank(from);
+        token.approve(address(this), type(uint256).max);
+
+        token.burn(from, burnAmount);
+
+        assertEq(token.totalSupply(), mintAmount - burnAmount);
+        assertEq(token.balanceOf(from), mintAmount - burnAmount);
+    }
+
+    function testApproveFuzz(address to, uint256 amount) public {
+        vm.assume(to != address(0) && to != address(token));
+
+        assertTrue(token.approve(to, amount));
+
+        assertEq(token.allowance(address(this), to), amount);
+    }
+
+    function testTransferFuzz(address to, uint256 amount) public {
+        vm.assume(to != address(0) && to != address(token));
+
+        token.mint(address(this), amount);
+
+        assertTrue(token.transfer(to, amount));
+
+        assertEq(token.totalSupply(), amount);
+        if (address(this) == to) {
+            assertEq(token.balanceOf(address(this)), amount);
+        } else {
+            assertEq(token.balanceOf(address(this)), 0);
+            assertEq(token.balanceOf(to), amount);
+        }
+    }
+
+    function testTransferFromFuzz(address to, uint256 allowance, uint256 amount) public {
+        vm.assume(to != address(0) && to != address(token));
+        amount = bound(amount, 0, allowance);
+
+        address from = address(0xABCD);
+        token.mint(from, amount);
+        vm.prank(from);
+        token.approve(address(this), allowance);
+
+        assertTrue(token.transferFrom(from, to, amount));
+
+        assertEq(token.totalSupply(), amount);
+
+        uint256 newAllowance = from == address(this) || allowance == type(uint256).max ? allowance : allowance - amount;
+        assertEq(token.allowance(from, address(this)), newAllowance);
+
+        if (from == to) {
+            assertEq(token.balanceOf(from), amount);
+        } else {
+            assertEq(token.balanceOf(from), 0);
+            assertEq(token.balanceOf(to), amount);
+        }
+    }
+
+    function testPermitFuzz(uint248 privateKey, address to, uint256 amount, uint256 deadline) public {
+        deadline = bound(deadline, block.timestamp, block.timestamp + type(uint80).max);
+        privateKey = uint248(bound(privateKey, 1, type(uint248).max));
```

Comment:
Can you explain why you've made these changes from solmate's version?

Reply:
Solmate's test suite is broken when it comes to error conditions.

Most of the tests using `testFail*` are false positives, so any failure in the test execution itself is swallowed, and the test passes regardless of the reason for failure.

When using `secp256k1` functions, there are some upper limits in the private key value:

> [FAIL. Reason: Private key must be less than 115792089237316195423570985008687907852837564279074904382605163141518161494337 (the secp256k1 curve order).

For that reason, whenever using `vm.address(privateKey)` with fuzz parameters, we need to restrict its type to `uint248`, because `uint256` can have numbers larger than the upper limit.